streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -424,23 +409,69 @@ public void evictWhile(final Supplier<Boolean> predicate,
         }
     }
 
+    @Override
+    public boolean hasKey(final K key) {
+        final Bytes serializedKey = Bytes.wrap(keySerde.serializer().serialize(changelogTopic, key));
+        return index.containsKey(serializedKey);
+    }
+
+    @Override
+    public ValueAndTimestamp<V> priorValueForBuffered(final K key) {
+        final Bytes serializedKey = Bytes.wrap(keySerde.serializer().serialize(changelogTopic, key));
+        final byte[] serializedValue = internalPriorValueForBuffered(serializedKey);
+
+        final V deserialize = valueSerde.innerSerde().deserializer().deserialize(
+            changelogTopic,
+            serializedValue
+        );
+
+        // it's unfortunately not possible to know this, unless we materialize the suppressed result, since our only
+        // knowledge of the prior value is what the upstream processor sends us as the "old value" when we first
+        // buffer something.
+        return ValueAndTimestamp.make(deserialize, RecordQueue.UNKNOWN);
```

Comment:
Can we use the `bufferValue.ContextualRecord.recordContext.timestamp` inside `internalPriorValueForBuffered` here?

Reply:
Unfortunately, that would be the timestamp of the "new value", but what we need is the timestamp of the "old value". Similar to what we discussed on the KIP-258 implementation, it's not super clear what the best surrogate value would be, but it seems better to guess at a lower-than-correct time, instead of a higher-than-correct time, so that we can at least let the other side of the join "win" and set the join result timestamp. But this is a short-term pragmatic decision.

I've been mulling over this, and I think I'd like to make a jira to allow forwarding the "old value" timestamp in the DSL implementation processors.

Reply:
Ack, that makes sense.